strings/base_collections_base.h

```diff
@@ -70,43 +140,66 @@ WINRT_EXPORT namespace winrt
 
             T Current() const
             {
-                if (m_current == m_end)
+                return m_owner->perform_shared([&]
                 {
-                    throw hresult_out_of_bounds();
-                }
+                    this->check_version(*m_owner);
 
-                if constexpr (!impl::is_key_value_pair<T>::value)
-                {
-                    return m_owner->unwrap_value(*m_current);
-                }
-                else
-                {
-                    return make<impl::key_value_pair<T>>(m_owner->unwrap_value(m_current->first), m_owner->unwrap_value(m_current->second));
-                }
+                    if (m_current == m_end)
+                    {
+                        throw hresult_out_of_bounds();
+                    }
+
+                    return current_value();
+                });
             }
 
-            bool HasCurrent() const noexcept
```

Comment:
Is losing noexcept here ok?  Previously this was nofailish, now there's an issue with version checking.

Reply:
I believe the version check is correct, and was missing from the old code. `m_current` and `m_end` can be invalidated by collection changes.

Reply:
To add on to Ryan's comment, the invalidation checking was previously done by `abi_enter` which is called before each function. Since this check needs to be synchronized it was moved here, which no longer made the function `noexcept`.